### PR TITLE
Create a simple tutorial for `Aligning`

### LINKED ISSUE #414: Write tutorials for `align` module
`graspy.align` is being added by #410. 

That PR will not have a tutorial. 

Need to add a notebook (or multiple notebooks) with simple and clear demonstration of functionality in that module.

### PR DESCRIPTION
**Pull request by Jingyan230**
**Reference Issues/PRs**
Fixes #414

**What does this implement/fix? Explain your changes.**
Add tutorial for Aligning

**Any other comments?**
Further improvements needed

### CODE REVIEW

File: docs/tutorial.rst [hunk 1]
```diff
@@ -81,3 +81,24 @@ including an introduction to the module, and how to utilize the seeding feature.
...
+   tutorials/subgraph/subgraph
+
+Aligning
+========
+The following tutorials shows how to align two seperate datasets with each other.
+
+.. toctree::
+   :maxdepth: 1
+
+   tutorials/matching/aligning
```
Comment (alyakin314) [inline docs/tutorial.rst:104]: aligning != matching; shouldn't go inside.

**Review by alyakin314 — CHANGES_REQUESTED** [review]

File: docs/tutorial.rst (hunk 1 above)
Comment (alyakin314) [reply docs/tutorial.rst:104]: as in shouldn't go inside of the matching folder

Comment (alyakin314) [reply docs/tutorial.rst:104]: or is this a misprint>

Comment (Jingyan230) [reply docs/tutorial.rst]: That's a misprint, sorry.

**Comment by alyakin314** [comment]
i'd say we need separate tutorials for each module. even the simple ones.

**Comment by bdpedigo** [comment]
> i'd say we need separate tutorials for each module. even the simple ones.

do you mean each class? separate notebook for sign flips, OP, SP?

**Comment by alyakin314** [comment]
> > i'd say we need separate tutorials for each module. even the simple ones.
> 
> do you mean each class? separate notebook for sign flips, OP, SP?

sorry, yes, obviously meant each class.

**Comment by bdpedigo** [comment]
> > > i'd say we need separate tutorials for each module. even the simple ones.
> > 
> > 
> > do you mean each class? separate notebook for sign flips, OP, SP?
> 
> sorry, yes, obviously meant each class.

I think me and jovo in class today disagree. was really nice to be able to compare performance of algos side by side. why did you want one for each @alyakin314? seems like each would be like 1 cell?

**Comment by alyakin314** [comment]
> > > > i'd say we need separate tutorials for each module. even the simple ones.
> > > 
> > > 
> > > do you mean each class? separate notebook for sign flips, OP, SP?
> > 
> > 
> > sorry, yes, obviously meant each class.
> 
> I think me and jovo in class today disagree. was really nice to be able to compare performance of algos side by side. why did you want one for each @alyakin314? seems like each would be like 1 cell?

the primary reason i thought this was that they have overlapping, but certainly not identical settings where they are appropriate: orthogonal requires matching, unlike the other two. sign flips only handle flips along the axes, unlike the other two which handle any rotations.

i guess you can force it to be one cell, though even without graphs examples, you can probably come up with multiple examples that highlight different points. for example, have a dataset where Y is really a rotated X, vs. when the distribution is rotated and the true solution is the points on top of each other. can also do more than two dimensions.

i am not saying that all of those need to be in the tutorial, mostly just bouncing around ideas.

**Comment by bdpedigo** [comment]
notes from office hours today
- start by making one notebook, we can split up later if it looks like too much
- add a section showing when sign flips will work
- add a section showing when seedless is helpful
   - this is when we don't have a matching between points, or there are different numbers of points 
   - try using multivariate normal distribution https://numpy.org/doc/stable/reference/random/generated/numpy.random.multivariate_normal.html

**Comment by kareef928** [comment]
> ## Peer review checklist
> ### PR itself (on GitHub)
> * [x] PR has a descriptive and succinct title.
> * [x] PR has a _brief_ description of what was changed.
> * [x] PR is addressing a change that has already been brought up in an issue. This will ensure that the change you propose is desired by the maintainers, and may provide a chance to discuss implementation details prior to PRing which will save everyone time.
> * [x] PR cites any related issues in the PR description and uses [closing keywords](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for any issues that should be closed as a result of the PR.
> * [x] PR does not have any extraneous file changes associated with it. One common example of an extraneous change is one that was already made in the target branch, but is showing up as part of this PR under `Files changed`.
> * [x] PR does not have any unresolved merge conflicts.
> * (NA) If implementing a major new feature or algorithm, a notebook demonstrating the proof of effectiveness is linked in the PR description. Depending on the feature, this kind of validation may take too long to run as a test or in a tutorial notebook.
> 
> ### Style
> * [x] Code follows Python variable naming conventions, one description is [here](https://visualgit.readthedocs.io/en/latest/pages/naming_convention.html). Most variables should be `snake_case` and classes should be defined as `UpperCamelCase`, for example.
> * [x] Code uses descriptive variable names throughout.
> * [x] Code has no commented out "junk code."
> * [ ] Line lengths are short (recommended <88 chars). This includes docstrings. Note:  [black](https://black.readthedocs.io/en/stable/) tries to shorten line lengths but sometimes it is unable to do so automatically, especially for docstrings.
_Couple of instances of plt.scatter(...) where the line length>88 chars_
> 
> ### Documentation
> * (NA) Any new public classes/functions have docstrings.
> * (NA) Any new public classes/functions have been added to the appropriate `.rst` file [here](https://github.com/microsoft/graspologic/tree/dev/docs/reference) to be rendered by `Sphinx`.
> * (NA) Any major new features (like a new algorithm) are accompanied by a succinct tutorial notebook. Or, if it is agreed upon with the maintainers that this is out of scope for the current PR, a new issue has been created specifying that we need a tutorial notebook for this functionality.
> * [x] Any new tutorial notebooks have been added to the appropriate folder [here](https://github.com/microsoft/graspologic/tree/dev/docs/tutorials) and included in the `.rst` file [here](https://github.com/microsoft/graspologic/blob/dev/docs/tutorial.rst) to be rendered by `Sphinx`.
> * [x] Modifications to the documentation render appropriately in the `Netlify` build.
> * (NA) References to parameters in the documentation are wrapped in double backticks (`` on either side of the parameter).
> * (NA) References to classes/function/documentation outside of this package use intersphinx to automatically link to the relevant documentation pages.
> 
> ### Testing
> * (NA) New public classes/functions are tested to ensure they achieve the desired output.
> * (NA) New public classes/functions are tested to ensure proper errors are thrown when invalid inputs are passed.

**Comment by Jingyan230** [comment]
Hi @alyakin314 , a new version of aligning tutorial with more descriptions is here. https://deploy-preview-573--graspy.netlify.app/tutorials/aligning/aligning.html
The same version in jupyter notebook here. https://nbviewer.jupyter.org/github/Jingyan230/notebook/blob/my-feature/aligning.ipynb

**Comment by bdpedigo** [comment]
@alyakin314 would you have time to give this a review (2/2)?

**Comment by alyakin314** [comment]
> @alyakin314 would you have time to give this a review (2/2)?

will try to do tomorrow morning be4 the class meeting (for both 1 and 2).

**Comment by bdpedigo** [comment]
> > @alyakin314 would you have time to give this a review (2/2)?
> 
> will try to do tomorrow morning be4 the class meeting (for both 1 and 2).

no huge rush. but thank you!!

**Review by alyakin314 — COMMENTED** [review]
not an approval for now, at least from me. a lot of style things, in particular all variable and module names need to be reformatted properly. shouldn't be too hard, as these are mostly search and replace. in addition to that, i think some writing should be a little more formal. address these comments, please, and i'll do another pass.

File: docs/tutorials/aligning/aligning.ipynb [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+    "%matplotlib inline"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Simple rotation example for each classes\n",
+    "\n",
+    "This part mainly explains how to use SignFlips, SeedlessProcrustes, and OrthogonalProcrustesFirst, with a simple rotation input example. First, let's build two simple apart datasets, where $Y$ is a randomly rotated group from $X$."
```
[the remark this answers is not in the document]
Comment (alyakin314) [reply docs/tutorials/aligning/aligning.ipynb:39]: and then fix throughout the tutorial.

File: docs/tutorial.rst [hunk 3]
```diff
@@ -93,3 +93,14 @@ The following tutorial demonstrates how to estimate the signal-subgraph of sampl
    :maxdepth: 1
 
    tutorials/subgraph/subgraph
+
+.. _aligning_tutorials:
+
+Aligning
```
[the remark this answers is not in the document]
Comment (bdpedigo) [reply docs/tutorial.rst:99]: agree 100% but is a separate issue IMO

File: docs/tutorials/aligning/aligning.ipynb [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+   ],
+   "source": [
+    "plt.subplots(1, 1, figsize=(6, 6))\n",
+    "plt.title(\"SignFlips, Simple rotation example\")\n",
+    "sns.color_palette(\"deep\")\n",
+    "plt.scatter(X[:,0], X[:,1], label=\"X\", marker='x')\n",
+    "plt.scatter(Y[:,0], Y[:,1], label=\"Y\", s=100, alpha=0.5)\n",
+    "plt.scatter(X_prime_SignFlips[:,0], X_prime_SignFlips[:,1], label=\"X@Q_SignFlips\", \n",
+    "            marker='$F$')\n",
+    "plt.tick_params(labelleft=False, labelbottom=False, bottom=False, left=False)\n",
```
[the remark this answers is not in the document]
Comment (bdpedigo) [reply docs/tutorials/aligning/aligning.ipynb:120]: i could see the case for leaving ticks for x=0 and y=0 but not much else

[the remark this answers is not in the document]
Comment (alyakin314) [reply docs/tutorials/aligning/aligning.ipynb:120]: sure, those are certainly most important ones. how do you feel about ticks at x=0, y=0 + grid? this should render as just x and y axis, in my understanding, which clearly specifies the origins and the axis that sign flips flip over.

File: docs/tutorial.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (alyakin314) [reply docs/tutorial.rst:99]: for sure.

File: docs/tutorials/aligning/aligning.ipynb [hunk 5]
```diff
@@ -0,0 +1,578 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Aligning\n",
+    "`Aligning` is a module for you to align two separated datasets together, for better comparison of the data. The main idea is to get a matrix $Q$ from the input data sets $X$ and $Y$, Then use $Q$ for aligning $X$ to where $Y$ is. The differences among classes in `Aligning` is the principle of how we get the orthogonal matrix $Q$ from inputs. For more information about the algorithm of these classes, see [reference](https://graspy.neurodata.io/reference/align.html)."
```
[the remark this answers is not in the document]
Comment (Jingyan230) [reply docs/tutorials/aligning/aligning.ipynb:8]: > Here and everywhere: the module is called `align` and can only be imported as that, so saying "among classes in `aligning`" is confusing.

In the Reference part, the module is named as `Aligning`, would that be confusing if the Tutorial name the module as `Align`...?

[the remark this answers is not in the document]
Comment (alyakin314) [reply docs/tutorials/aligning/aligning.ipynb:8]: the names of our "chapters" are a question for a different day (see below). but it certainly should be called align when to referred as a module.

[the remark this answers is not in the document]
Comment (Jingyan230) [reply docs/tutorials/aligning/aligning.ipynb:8]: OK thanks.

File: docs/tutorials/aligning/aligning.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jingyan230) [reply docs/tutorials/aligning/aligning.ipynb:39]: I didn't find a special text style for the names of different classes in other tutorials, please correct me if I missed it...

**Comment by Jingyan230** [comment]
Hi, @alyakin314 . I made some changes according to the comments above, please tell me if I missed something or further improvements are needed.

**Comment by alyakin314** [comment]
could you plot just X and Y before the Sign Flips section?

File: docs/tutorials/aligning/aligning.ipynb [hunk 6]
```diff
@@ -0,0 +1,583 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Aligning\n",
+    "`Align` is a module for aligning two separated datasets together, in order to have better comparison of the data. The main idea is to get a matrix $Q$ from the input data sets $X$ and $Y$, Then use $Q$ for aligning $X$ to where $Y$ is. The differences among classes in `Align` is the principle of how we get the orthogonal matrix $Q$ from inputs. For more information about the algorithm of these classes, see [reference](https://graspy.neurodata.io/reference/align.html)."
```
Comment (alyakin314) [inline docs/tutorials/aligning/aligning.ipynb:8]: do not capitalize `align` when it's within ticks, please.

**Comment by alyakin314** [comment]
im going to take a careful review today

**Comment by alyakin314** [comment]
also, could you remove all ticks but X = 0 and Y = 0. im indifferent to leaving the rest, but it would make Ben happy, and we like Ben happy.

**Comment by alyakin314** [comment]
okay. i've wanted to edit some grammar/writing but got a little to into it and also swapped one example for another and wrote out more descriptions in various places.

**Comment by alyakin314** [comment]
i should've called the assignment matrix P, and not W. could you change it to that, @Jingyan230 ?

**Comment by alyakin314** [comment]
could you also add a heatmap of Z in the very beginning, similar to how i provided a heatmap of Q_. (and also change the name from Z to Q to avoid further confusion).

**Comment by alyakin314** [comment]
couple comments from Youngser, @Jingyan230 

- line 9: two "which"s.
- line below Fig 3: "third dimension" => "third quadrant".

**Comment by alyakin314** [comment]
change h to heuristic in the second to last cell; remove the last cell.

**Comment by alyakin314** [comment]
could you add a throwaway comment saying that if two graphs are embedded using omni - they need not to be aligned

**Review by alyakin314 — APPROVED** [review]
sent it to carey, and he liked it. im going to approve this subject to fixing the issues mentioned above and Ben also approving it

**Comment by Jingyan230** [comment]
Hi @alyakin314 , I changed the plot in the Shuffled Example from X to X_shuffled, I think that makes more sense. If you have different consideration for ploting X instead of X_shuffled, please tell me. I can change it back.

**Comment by alyakin314** [comment]
> Hi @alyakin314 , I changed the plot in the Shuffled Example from X to X_shuffled, I think that makes more sense. If you have different consideration for ploting X instead of X_shuffled, please tell me. I can change it back.

plot's the same tho? you can leave it as is, obviously.

**Comment by bdpedigo** [comment]
@Jingyan230 @alyakin314 : I like the updated notebook and I think it's basically good to go. Just one comment throughout - everywhere you use quasi-latex notation can become actual latex rendered by matplotlib. 

An example from cell [4]: change to
```
ax.scatter(X[:,0], X[:,1], label=r"$X$", marker='x')
ax.scatter(Y[:,0], Y[:,1], label=r"$Y$", s=100, alpha=0.5)
```

An example from cell [7]: change to 
```
ax.scatter(X_prime_SF[:,0],
           X_prime_SF[:,1],
           label=r"$X_{SF}$",
           marker='$F$')
```

An example from cell [21]: change to 
```
heatmap(aligner_SF_shuffled.Q_, title=r'$Q_{SF}$', vmin=-1, vmax=1, ax=ax2)
```

Make sense? I think it just makes it easier to read and match the math formatting in the text. @Jingyan230 if you could make these small changes throughout the notebook, I think we are ready to merge. Basically you just have to put the `r` in front of the appropriate strings, and add $$ to make the appropriate places render as math/latex in matplotlib.

**Comment by alyakin314** [comment]
^ that will look much better indeed.

**Comment by Jingyan230** [comment]
@bdpedigo The text format problem is solved. It's ready for checking. And I add a line in the last cell to print the mean of X_SF.

**Review by bdpedigo — APPROVED** [review]
@jingyan230 looks good, thank you and thanks @alyakin314!

**Pull request merged**